libretro-common/include/libretro.h

```diff
@@ -3030,6 +3052,89 @@ struct retro_disk_control_ext_callback
    retro_get_image_label_t get_image_label;     /* Optional - may be NULL */
 };
 
+/* Callbacks for RETRO_ENVIRONMENT_SET_NETPACKET_INTERFACE.
+ * A core can set it if sending and receiving custom network packets
+ * during a multiplayer session is desired.
+ */
+
+/* Used by the core to send a packet to one or more connected players.
+ * A single packet sent via this interface can contain up to 64kb of data.
+ *
+ * If the ready callback has indicated the local player to be the host:
+ *  - The broadcast flag can be set to true to send to multiple connected clients
+ *  - On a broadcast, the client_id argument indicates 1 client NOT to send the packet to
+ *  - Otherwise, the client_id argument indicates a single client to send the packet to
+ * If the local player is a client connected to a host:
+ *  - The broadcast flag is ignored
+ *  - The client_id argument must be set to 0
+ *
+ * This function is not guaranteed to be thread-safe and must be called during
+ * retro_run or any of the netpacket callbacks passed with this interface.
+ */
+typedef void (RETRO_CALLCONV *retro_netpacket_send_t)(const void* buf, size_t len, uint16_t client_id, bool broadcast);
+
+/* Called by the frontend to signify that a multiplayer session has started.
+ * If client_id is 0 the local player is the host of the session and at this
+ * point no other player has connected yet.
+ *
+ * If client_id is > 0 the local player is a client connected to a host and
+ * at this point is already fully connected to the host.
+ *
+ * The core will have to store the retro_netpacket_send_t function pointer
+ * passed here and use it whenever it wants to send a packet. That send
+ * function pointer is valid until the frontend calls retro_netpacket_stop_t.
+ */
+typedef void (RETRO_CALLCONV *retro_netpacket_start_t)(uint16_t client_id, retro_netpacket_send_t send_fn);
+
+/* Called by the frontend when a new packet arrives which has been sent from
+ * a connected client or the host with retro_netpacket_send_t.
+ * The client_id argument indicates who has sent the packet. On the host side
+ * this will always be > 0 (coming from a connected client).
+ * On a client connected to the host it is always 0 (coming from the host).
+ * Packets sent with this interface arrive at this callback in a reliable
+ * manner, meaning in the same order they were sent and without packet loss.
+ */
+typedef void (RETRO_CALLCONV *retro_netpacket_receive_t)(const void* buf, size_t len, uint16_t client_id);
```

Comment:
Shouldn't you specify requirements for the transport layer somewhere? Cores don't need to care, but frontends will. "Reliable" could mean TCP, WebSockets, something custom on top of UDP...

Reply:
I agree, ideally `retro_netpacket_send_t` would have something like a `enum retro_netpacket_flag flag` argument which could be set to various modes of sending (i.e. `reliable`, `unreliable`, `unsequenced`). But at least in the implementation I can contribute it will be limited to what RetroArch netplay currently can do, which is only `reliable` as it goes over TCP.

Maybe we add the flag for now for future expansion but only have `reliable` as an option? Or nothing and have reliable be the default.

Reply:
> Maybe we add the flag for now for future expansion but only have reliable as an option?

I like this idea, but I still think you should explicitly document how the protocol (at the application level) works. You might say "TCP", and that's fine, but what can other frontends expect in the payload? Just the buffer? Some preceding metadata?

Reply:
As the interface stands now it is all transparent to the core. The core sends 10 bytes with `retro_netpacket_send_t`, the other side will receive that 10 bytes with `retro_netpacket_receive_t`. By design the core shouldn't care what communication was used to send the packet. It could be pigeon post :-)
This is what I meant with this part of the spec:
```
* The frontend will take care of connecting players together.
* The core only needs to send the actual data as needed for the
* emulation while handshake and connection management happens in
* the background.
```
A frontend other than RetroArch could use websockets to transmit the data, or basing it on a library like [ENET](http://enet.bespin.org/). In the end it doesn't matter if it's TCP or UDP or something else. This is not meant to connect frontends together. If another frontend wants to support netplay with RetroArch, then it needs to follow the specs of netplay in RetroArch - which are not part of libretro.h.

Reply:
In the new commit c0f85f8 I added a `flags` argument to `retro_netpacket_send_t` with the following options:
```c
/* Netpacket flags for retro_netpacket_send_t */
#define RETRO_NETPACKET_UNRELIABLE  0        /* Packet to be sent unreliable, depending on network quality it might not arrive. */
#define RETRO_NETPACKET_RELIABLE    (1 << 0) /* Reliable packets are guaranteed to arrive at the target in the order they were send. */
#define RETRO_NETPACKET_UNSEQUENCED (1 << 1) /* Packet will not be sequenced with other packets and may arrive out of order. Cannot be set on reliable packets. */
```

Also I extended the comment for `retro_netpacket_send_t` with this:
```c
 * A frontend must support sending of reliable packets (RETRO_NETPACKET_RELIABLE).
 * Unreliable packets might not be supported by the frontend but the flags can
 * still be specified, reliable transmission will be used instead.
```
Because NetPlay in RetroArch can only send packets reliable. Other frontends may use communication over something like UDP which would enable the other reliability modes.